### PR TITLE
ARGO-375 Add Authentication to Messaging API

### PR DESCRIPTION
**Pull request by kaggis**
### Goals

- [x] Add Authentication mechanism 
- [x] Add also authorization (role based system)

### Authentication
In each request a api key token is provided as url parameter (spec)
e.g. GET `/v1/project/PROJECT101/topics?key=T0K3N`

Authentication mechanism checks in store if the project/key combination corresponds to an existing
user. 

Authentication is performed through a WrapAuthentication wrapper that can be appended to handler chain on demand. A configuration parameter is used `use_authentication:true` to enable/disable the use of authentication on the API

### Authorization

After the user has been authenticated, there is an authorization process based on roles to see if the user has access to the resource. In the store there are entries in the roles collection like the following
```json
{ "resource": "topics:list", "roles": ["admin","reader","publisher"] }
```
After authentication the role list of a user is passed to the authorization module and the latter decides if the resource is accessible or not. A configuration parameter is used `use_authorization:true` to enable/disable role authorization on the API

_Note_: thus the api can be started with authentication:on/off and authorization:on/off (provided authentication is also on)

## Implementations

- [x] Add Checks for project exists in store. Use for verification in handlers (topic:publish, subscription:pull)
- [x] Add Checks for project/topic , project/subscription exists. Use for verification in handlers (topic:publish, subscription:pull)
- [x] Update unittests
- [x] Add auth package
- [x] Implement Authentication wrapper to check if a token corresponds to a user of project
- [x] Implement Authorization wrapper to check if resource is accessible based on role given

- [x] swagger updated

### CODE REVIEW

**Comment by kaggis** [comment]
please review
/cc @pkoro @dpavlos

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

**Comment by themiszamani** [comment]
test this please

File: handlers_test.go [hunk 1]
```diff
@@ -319,12 +319,67 @@ func (suite *HandlerTestSuite) TestPublishError() {
...
 
+func (suite *HandlerTestSuite) TestPublishNoTopic() {
+
+	postJSON := `{
+  "messages": [
+    {
+      "attributes": [
+        {
+          "key": "foo",
+          "valu2RlZA=="
```
Comment (pkoro) [inline handlers_test.go]: I am not sure about this line. Should it be smth like `"value": "bar"`?

File: config.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "broker_host":"localhost:9092",
+  "broker_host":"as01.hadoop:9092",
```
Comment (pkoro) [inline config.json]: Maybe we should hide this. I understand this aids the development process so my suggestion would be to add `config.json` in `.gitignore` and just provide a `config.json.template` in the codebase as an example .

Comment (kaggis) [inline config.json]: This should be localhost, will change it

File: auth/auth_test.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	store := stores.NewMockStore("mockhost", "mockbase")
+	suite.Equal([]string{"admin", "member"}, Authenticate("ARGO", "S3CR3T", store))
+	suite.Equal([]string{}, Authenticate("ARGO", "falseSecret", store))
+
+	suite.Equal(true, Authorize("topics:list_all", []string{"admin"}, store))
+	suite.Equal(true, Authorize("topics:list_all", []string{"admin", "reader"}, store))
+	suite.Equal(true, Authorize("topics:list_all", []string{"admin", "foo"}, store))
+	suite.Equal(false, Authorize("topics:list_all", []string{"foo"}, store))
+	suite.Equal(false, Authorize("topics:publish", []string{"reader"}, store))
+	suite.Equal(true, Authorize("topics:list_all", []string{"admin"}, store))
```
Comment (pkoro) [inline auth/auth_test.go]: This line seems redundant as it seems the same with L20.

File: stores/store_test.go [hunk 4]
```diff
@@ -16,17 +16,35 @@ func (suite *StoreTestSuite) TestMockStore() {
...
+	suite.Equal([]string{"admin", "member"}, store.GetUserRoles("ARGO", "S3CR3T"))
+	suite.Equal([]string{}, store.GetUserRoles("ARGO", "SecretKey"))
+
+	// Test roles
+	suite.Equal(true, store.HasResourceRoles("topics:list_all", []string{"admin"}))
+	suite.Equal(true, store.HasResourceRoles("topics:list_all", []string{"admin", "reader"}))
+	suite.Equal(true, store.HasResourceRoles("topics:list_all", []string{"admin", "foo"}))
+	suite.Equal(false, store.HasResourceRoles("topics:list_all", []string{"foo"}))
+	suite.Equal(false, store.HasResourceRoles("topics:publish", []string{"reader"}))
+	suite.Equal(true, store.HasResourceRoles("topics:list_all", []string{"admin"}))
```
Comment (pkoro) [inline stores/store_test.go]: Same comment as previous for this line.

**Comment by pkoro** [comment]
+1 other than the two comments made inline.

File: auth/auth_test.go (hunk 3 above)
Comment (kaggis) [inline auth/auth_test.go]: Indeed, removed

**Comment by pkoro** [comment]
+1

**Pull request merged**